### ISSUE TITLE
Where should Lattices be?

### ISSUE THREAD

**Issue by HuStmpHrrr** [agda/agda-stdlib#439]
There are a few implementations of semilattices/lattices:

https://github.com/agda/agda-stdlib/blob/master/src/Algebra/Structures.agda#L410
https://github.com/agda/agda-stdlib/blob/master/src/Relation/Binary/Lattice.agda

The ones implemented in `Relation.Binary.*` are more general, and one can show that the `Relation.Binary.Lattice` induces ones from `Algebra.Structures`. However, I do not understand why semilattices/lattices are defined in `Relation.Binary`? I tend to understand them to be algebras, but with order theoretic settings. Shall be move them to `Algebra`? Since order theory is rich, we can probably put them in `Algebra.OrderTheory.*`. What do you think?

**Comment by JacquesCarette** [comment]
I agree that `Algebra.OrderTheory.*` would be better than `Relation.Binary.Lattice`.

**Comment by HuStmpHrrr** [comment]
Then the next question is where should the ones from `Alegbra.Structures` go if we decide to move the definitions to `Algebra.OrderTheory`?

**Comment by WolframKahl** [comment]
@JacquesCarette, partial orders are relational structures, and as such not algebras.

@HuStmpHrrr, the binary-suprema-and-infima characterisation of lattices as special partial orders
is (in general) actually equivalent to the algebraic definition.

@JacquesCarette, but it is not of algebraic nature, so `Algebra.OrderTheory` sounds wrong.

Just `OrderTheory.*`? Or even just `Order.*`?

**Comment by HuStmpHrrr** [comment]
@WolframKahl Is algebra necessarily equational? I am getting confused. The version of Heyting algebra I learned is directly defined over preorder/partial order, and equivalence is not mentioned. From there I do think lattice theory on partial order is an algebra.

**Comment by HuStmpHrrr** [comment]
@WolframKahl thanks. understood. indeed in that sense order theory is out of algebraic theories. Do you think it's better to leave them as relation, or put them in `Order.*`? I still find surprising to see lattices as binary relation.

**Comment by MatthewDaggitt** [comment]
Hmm my opinion (an admittedly fairly uninformed one!) is that lattices are algebras as they have operators. I'd also vote for `Algebra.OrderTheory.*` over `OrderTheory.*` or `Order.*` because both of the latter sound like they do actually belong in `Relation.Binary`.

**Comment by WolframKahl** [comment]
Lattices **defined via two associative, commutative and idempotent operators** satisfying the (equational) absorption laws are algebras.

Lattices **defined as partial orders** where binary suprema and infima exist (via quantified predicate logic formulae) are not algebras.

**Comment by HuStmpHrrr** [comment]
I want to add a few more points on why I think these theories should at least be pulled out `Relation.Binary.*`:

1. These theories all contain an equivalence, so they possess algebras. For example, infimum and supremum are commutative semigroup, and commutative monoid if they have bounds. However, I am not sure in which way I am supposed to claim that these semilattices have algebraic structures, because these structures should not reside in `Relation.Binary.*` already.

2. There is a huge intersection between algebraic definitions and order theoretic ones. In my opinion, we should probably formalize them using order theory, and just induce corresponding algebraic structures. 

Regarding the topic of algebraic theories, one thing I cannot wrap my head around with, is why Heyting algebra and Boolean algebra are algebras? Especially HA, I often see exponential is via order theory and the definition itself is also very elegant. Is that a historical abuse of terminology?

**Comment by JacquesCarette** [comment]
@HuStmpHrrr A Boolean Algebra is an Algebra because they have a presentation which is purely algebraic. There are other modes of presentation which are indeed more order-theoretic.

Interestingly, I cannot recall a purely algebraic definition for Heyting "algebras"!

This is not entirely surprising. Mathematics severely abuses the word 'algebra'. The fact that in some instances, it can be made formal at all, is the more surprising thing!

**Comment by MatthewDaggitt** [comment]
Removing this from the v1.0 milestone, as this doesn't feel like a pressing issue.

**Comment by jamesmckinna** [comment]
Six years on, we seem to have stabilised our design decisions into `Relation.Binary.Lattice.Structures` (order-theoretic) and `Algebra.Lattice.Structures` (algebraic), so I'm guessing this is OK to close now?
